sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,219 @@
+           [ 1., 1., 1., 0.],
+           [ 2., 2., 2., 1.],
+           [ 2., 2., 2., 2.]])
+
+    Categorical features can be specified in the constructor.
+
+    >>> dis = KBinsDiscretizer(n_bins=3, categorical_features=[1])
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+
+    Discretization will ignore categorical features.
+    >>> dis.transform(X) # doctest: +SKIP
+    array([[ 0., 1., 0., 0.],
+           [ 1., 2., 1., 0.],
+           [ 2., 3., 2., 1.],
+           [ 2., 4., 2., 2.]])
+    """
+
+    def __init__(self, n_bins=2, categorical_features=None):
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+    def fit(self, X, y=None):
+        """Fits the estimator.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            Data to be discretized.
+
+        Returns
+        -------
+        self
+        """
+        X = check_array(X, dtype=float, ensure_2d=False)
+
+        if X.ndim == 1:
+            raise ValueError("Reshape your data.")
+
+        n_features = X.shape[1]
+        self._check_categorical_features(self.categorical_features, X.shape[1])
+        self._check_n_bins(self.n_bins, n_features)
+
+        if self.categorical_features:
+            self.categorical_features_ = np.asarray(self.categorical_features)
+        else:
+            self.categorical_features_ = np.array([], dtype='int64')
+
+        # Manipulating min and max so categorical data is not affected by
+        # transformation.
+        min = np.min(X, axis=0)
+        max = np.max(X, axis=0)
+
+        min[self.categorical_features_] = 0
+        max[self.categorical_features_] = 1
+
+        self.min_ = min
+        self.ptp_ = max - min
+
+        # Set n_bins_ so categorical data can be ignored by transformation.
+        if isinstance(self.n_bins, numbers.Number):
+            n_bins = np.ones(n_features) * self.n_bins
+        else:
+            n_bins = check_array(self.n_bins, dtype=int, ensure_2d=False,
+                                 copy=True)
+
+        n_bins[self.categorical_features_] = 1
+        self.n_bins_ = n_bins
+
+        # Clipping features ensures outliers are within boundary after
+        # transformation, but also ensures categorical features can be ignored.
+        numeric_features = np.delete(np.arange(n_features),
+                                     self.categorical_features_)
+        clip_min = np.repeat(-np.inf, n_features)
+        clip_min[numeric_features] = 0
+
+        clip_max = np.repeat(np.inf, n_features)
+        clip_max[numeric_features] = self.n_bins_[numeric_features] - 1
+
+        self._clip_min = clip_min
+        self._clip_max = clip_max
+
+        same_min_max = np.where(self.ptp_ == 0)[0]
+        if len(same_min_max) > 0:
+            import warnings
+            warnings.warn("Fitted X contained continuous features at indices "
```

Comment:
I'd like to know where the user would hit a similar warning. As far as I know, when users use this discretizer, this is the only warning users would receive when fitting data. 

The warning adds some noise to the code, but I feel that it does have a purpose.